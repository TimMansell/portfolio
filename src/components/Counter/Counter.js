import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

export const Counter = ({ begin, end }) => {
  const [counter, setCounter] = useState(begin);
  const [randomCountTo] = useState(
    Math.floor(Math.random() * (end - begin + 1) + begin) + 1
  );

  useEffect(() => {
    if (counter <= randomCountTo) {
      const timer = setInterval(() => {
        setCounter(counter + 1);
      }, 1);

      return () => clearInterval(timer);
    }
  });

  return <>{counter}+</>;
};

Counter.propTypes = {
  end: PropTypes.number.isRequired,
  begin: PropTypes.number.isRequired,
};

export default Counter;
