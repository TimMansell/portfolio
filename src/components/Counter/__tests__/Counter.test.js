import React from 'react';
import { render, act } from '@testing-library/react';
import '@testing-library/jest-dom';

import Counter from '../Counter';

const props = {
  begin: 0,
  end: 3,
};

describe('Counter', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  it('should render my component', () => {
    render(<Counter {...props} />);
  });

  it('should match snapshot', () => {
    const { asFragment } = render(<Counter {...props} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should start at correct number', async () => {
    const { getByText } = render(<Counter {...props} />);

    expect(getByText(props.begin)).toBeInTheDocument();
  });

  it('should end at correct number', async () => {
    const { getByText } = render(<Counter {...props} />);

    act(() => jest.advanceTimersByTime(9000));

    expect(getByText(props.end)).toBeInTheDocument();
  });
});
