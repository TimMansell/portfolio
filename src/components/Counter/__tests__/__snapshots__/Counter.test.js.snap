// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter should match snapshot 1`] = `
<DocumentFragment>
  0
</DocumentFragment>
`;
